### PR TITLE
Homematic: Fixed scheduled pool rejection on reconnect

### PR DESCRIPTION
**Pull request by gerrieg**
Signed-off-by: Gerhard Riegler <gerhard.riegler@gmail.com>

see https://github.com/openhab/openhab2-addons/issues/967 and https://github.com/openhab/openhab2-addons/issues/1023

### CODE REVIEW

**Comment by kaikreuzer** [comment]
The correct fix would actually be to call `ThreadPoolManager.getScheduledPool("homematicGateway")`at all places where you currently use `scheduler` - i.e. don't keep the instance anywhere in your code, but always ask the ThreadPoolManager for the currently valid instance of the pool.

**Comment by gerrieg** [comment]
You told me in https://github.com/openhab/openhab2-addons/issues/967#issuecomment-222360833, i should use the default scheduler from BaseThingHandler ;-)

But no problem, i changed it again and with the ThreadPoolFactory, the internal package is now completely decoupled from OH.

**Comment by kaikreuzer** [comment]
> You told me in #967 (comment), i should use the default scheduler from BaseThingHandler ;-)

That's right - but I was only thinking along the lines of using it within the handler code. You seem to reach i through to many other service instances that potentially have their own lifecycle (and thus can get out of sync with the the scheduler instance).

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/OpenHabThreadPoolFactory.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+package org.openhab.binding.homematic.handler;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import org.eclipse.smarthome.core.common.ThreadPoolManager;
+import org.openhab.binding.homematic.internal.common.ThreadPoolFactory;
+
+/**
+ * OpenHab implementation of the Homematic ThreadPoolFactory.
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/OpenHabThreadPoolFactory.java]: what is this class good for? I don't really understand...

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/ThreadPoolFactory.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Factory to get thread pools from the core runtime.
+ *
+ * @author Gerhard Riegler - Initial contribution
+ */
+public interface ThreadPoolFactory {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/common/ThreadPoolFactory.java]: And what is this good for? I do not understand either...

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/AbstractHomematicGateway.java [hunk 3]
```diff
@@ -80,8 +79,7 @@
     private long lastEventTime = System.currentTimeMillis();
     private DelayedExecuter delayedExecutor = new DelayedExecuter();
     private Set<HmDatapointInfo> echoEvents = Collections.synchronizedSet(new HashSet<HmDatapointInfo>());
-    // FIXME: dependency to openHab
-    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool("homematicGateway");
+    private ScheduledExecutorService scheduler;
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/AbstractHomematicGateway.java]: In https://github.com/openhab/openhab2-addons/pull/1078#issuecomment-229645333 I meant that you should not have a field for the scheduler at all.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/AbstractHomematicGateway.java [hunk 4]
```diff
@@ -117,6 +115,8 @@ public AbstractHomematicGateway(String id, HomematicConfig config, HomematicGate
     public void initialize() throws IOException {
         logger.debug("Initializing gateway with id '{}'", id);
 
+        scheduler = config.getThreadPoolFactory().getScheduledPool("homematicGateway");
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/communicator/AbstractHomematicGateway.java]: You won't need this.
Whereever in your code you want to schedule a job, simply use

```
ThreadPoolManager.getScheduledPool("homematicGateway");
```

**Comment by gerrieg** [comment]
ahaaa ... i completely misunderstood you, sorry. Now i think it's correct.

**Comment by kaikreuzer** [comment]
Aaaah, that looks so much better :-)

Thanks!

**Comment by luci84tm** [comment]
Sorry for the dummy question: How to get the latest Homematic addon with this issue fixed ?
Do I need Eclipse and compile the code into a JAR or?
(sorry I'm not a java developer)
Thanks allot!

**Pull request merged**